cql3/statements/ks_prop_defs.cc

```diff
@@ -74,20 +75,64 @@ std::optional<sstring> ks_prop_defs::get_replication_strategy_class() const {
     return _strategy_class;
 }
 
-lw_shared_ptr<keyspace_metadata> ks_prop_defs::as_ks_metadata(sstring ks_name) {
+std::map<sstring, sstring> ks_prop_defs::prepare_options(const sstring& strategy_class, const std::map<sstring, sstring>& old_options = {}) {
     auto options = get_replication_options();
     options.erase(REPLICATION_STRATEGY_CLASS_KEY);
-    return keyspace_metadata::new_keyspace(ks_name, get_replication_strategy_class().value(), options, get_boolean(KW_DURABLE_WRITES, true));
+
+    if (strategy_class != "NetworkTopologyStrategy") {
+        return options;
+    }
+
+    // For users' convenience, expand the 'replication_factor' option into a replication factor for each DC.
+    // If the user simply switches from another strategy without providing any options,
+    // but the other strategy used the 'replication_factor' option, it will also be expanded.
+    // See issue CASSANDRA-14303.
+
+    sstring rf;
+    auto it = options.find(REPLICATION_FACTOR_KEY);
+    if (it != options.end()) {
+        rf = it->second;
+        options.erase(it);
+    } else if (options.empty()) {
+        auto it = old_options.find(REPLICATION_FACTOR_KEY);
+        if (it != old_options.end()) {
+            rf = it->second;
+        }
+    }
```

Comment:
Why do you touch old_options here? Wouldn't the have their REPLICATION_FACTOR_KEY erased?

Reply:
Only in the case when the strategy class was previously NetworkTopologyStrategy would the old options have had their REPLICATION_FACTOR_KEY erased.

This comment explains it:
```
// If the user simply switches from another strategy without providing any options,
// but the other strategy used the 'replication_factor' option, it will also be expanded.
```
So if, for example, the previous strategy was SimpleStrategy, and we switch directly to NetworkTopologyStrategy using an ALTER statement without any other options than the class (`else if (options.empty())`, 'class' was erased), replication_factor will be expanded. The tests check for this case.

There was a discussion in the Cassandra issue whether this is the correct approach, and they decided to go with it:
https://issues.apache.org/jira/browse/CASSANDRA-14303?focusedCommentId=16713793&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16713793

Reply:
Added more comments to explain.